engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,201 @@
+# PLAN-0002: Ethereum Tracing Cache
+
+<dl>
+  <dt>Author</dt>
+  <dd>Zachary Burns</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0002-ethereum-tracing-cache.md">RFC-0002 Ethereum Tracing Cache</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/9">https://github.com/graphprotocol/rfcs/pull/9</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-20</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Summary
+
+Implements RFC-0002: Ethereum Tracing Cache
+
+## Implementation
+
+These changes happen within or near `ethereum_adapter.rs`, `store.rs` and `db_schema.rs`.
+
+### Limitations
+The problem of reorg turns out to be a particularly tricky one for the cache, mostly due to ranges of blocks being requested rather than individual hashes. To sidestep this problem, only blocks that are older than the reorg threshold will be eligible for caching.
+
+Additionally, there are some subgraphs which may require traces from all or a substantial number of blocks and don't make effective use of filtering. In particular, subgraphs which specify a call handler without a contract address fall into this category. In order to prevent the cache from bloating, any use of Ethereum traces which does not filter on a contract address will bypass the cache.
+
+### EthereumTraceCache
+
+The implementation introduces the following trait, which is implemented primarily by `Store`.
+
+```rust
+use std::ops::RangeToInclusive;
+struct TracesInRange {
+    range: RangeToInclusive<u64>,
+    traces: Vec<Trace>,
+}
+
+pub trait EthereumTraceCache: Send + Sync + 'static {
+    /// Attempts to retrieve traces from the cache. Returns ranges which were retrieved.
+    /// The results may not cover the entire range of blocks. It is up to the caller to decide
+    /// what to do with ranges of blocks that are not cached.
+    fn traces_for_blocks(contract_address: Option<H160>, blocks: Range<u64>
+        ) -> Box<dyn Future<Output=Result<Vec<TracesInRange>, Error>>>;
+    fn add(contract_address: Option<H160> traces: Vec<TracesInRange>);
+}
+```
+
+#### Block schema
+
+Each cached block will exist as its own row in the database in an `eth_traces_cache` table.
+
+```rust
+eth_traces_cache(id) {
+  id -> Integer,
+  network -> Text,
```

Comment:
For efficiency, we could limit the length of network names to e.g. 32 or 40 characters.

Reply:
@lutter Just said on slack that one should never limit the length of a VarChar. Perhaps he can elaborate.

One approach that works well in games is just to use a hash for any string that you want to use for compares/lookups. The collision probability for 20 networks with a 64 bit integer is 1 in 10^17 https://preshing.com/20110504/hash-collision-probabilities/ (There is a handy table at the end of this article)

I've applied this to hashmap structures and a few other places that required very fast text search based on eg: https://ourmachinery.com/post/minimalist-container-library-in-c-part-2/ In general, fixed size values and ints will be faster than strings.

Reply:
Really the most efficient thing, even more efficient than Postgres 11 hash table partitioning, is to just have different tables per network. I'm not sure what capabilities, if any, we have for creating tables on the fly though.

Reply:
We do create tables dynamically for subgraphs (David can share more details). It's definitely not hard. For a network like the Ethereum mainnet, we could create tables prefixed with  `ethereum_mainnet_`. I'm not sure it's worth doing here (yet).

As for hashing to e.g. a 64 bit integer... even the smallest probability makes me uneasy because under no circumstances can we return traces for the wrong network. I can see it working wonders in combination with collision detection in a hash map/set though. It wouldn't be hard to maintain a table that maps network names to a unique number that we then use for the network column. 

Perhaps we can start without all this though, just use the network name and once we have 
everything implemented, we can refine it? What do you think?

Reply:
I'm all for getting this out the door.